Zad5

zimi-pl opened the pull request.

sm000k committed:
Zad dzialajaca potega 10

sm000k committed:
Zad3 dopieszczamy kod pierwsza iteracja

sm000k committed:
Zad4 reversing strings

sm000k committed:
Zad4 commit poprawnego 1-testu prymitywny przykład

sm000k committed:
Zad4 commit poprawnego 2-test poprawny z podstawowym wsadem leetcode

sm000k committed:
Zad4 commit dodanie 3 i 4 testu 3 poszedł bez problemów
z czwartym był
jak masz tablice stringów:
{"alfa","beta","",""}
To ciężko określić wartość albo zawartość komórek o treści "". Nie jest to null nie jest to "", ani tym bardziej " ". Przynajmniej mi się nie udało od razu znaleźć wyjścia z problemu

sm000k committed:
Zad4 commit zadanie rozwiązane tak że sprawdzam czy długość stringa w tablicy jest większy od zera. Jest to potrzebne bo split usuwa spacje ze stringa ale tworzy tablice z komórkami bez zawartości, których wartość jest dla mnie nieznana, ale za pomocą .lenth() jestem w stanie powiedzieć czy komorka jest pusta, czy nie.

sm000k committed:
Zad5 stworzenie pierwszego testu

sm000k committed:
Zad5 sprawdzenie rozwiązania typu fuzzing z typem int.

sm000k committed:
Zad 4 poprawienie wielu rzeczy dostosowanie się do uwag Zimiego i kogoś jeszcze

sm000k committed:
Zad 4 poprawienie
If (splitedS[i].length() > 0)
na
if (!splitedS[i].equals(""))

usuniecie zmiennych o nieczytelnych nazwach

sm000k committed:
Zad 4
poprawiono nazwy zmiennych dodano obsługę przypadku brzegowego: null

sm000k committed:
Zad 4
dodano kolejny warunek brzegowy: s=""
usunięto zawartość metody main. Ok, test driven development, czyli bazujemy na danych testowych z testów i tylko stamtąd.

sm000k committed:
Zad 4
Dodanie Testu i obsługa przypadku dla podanej wartości równej ZERO.

sm000k committed:
Zad 5
Dodanie moich wypocin przy ciągu fibbonaciego. Wrzucam jako dowód włożonej pracy. Nie polecam sprawdzać.

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad4.java:
```diff
@@ -0,0 +1,76 @@
+        return reversedString = reversedString.substring(1, reversedString.length());
+    }
+
+    public static void main(String[] args) {
```
usuń

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad4.java:
```diff
@@ -0,0 +1,76 @@
+
+    public static void main(String[] args) {
+    }
+    /*
```
usuń zakomentowany kod

zimi-pl replied:
chyba że jest dla Ciebie jakoś ważny to zrób z tego niezakomentowany działający kod... np. testy, albo alternatywne rozwiązanie

zimi-pl replied:
dotyczy to linii do 75

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+            sumary = 0;
+            dummyPathString = "";
+        }
+//        for (String iter : treeSet) {
```
usuń zakomentowany kod

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+            }
+            if (sumary == n) {
+                treeSet.add(dummyPathString);
+//                allPasscount++;
```
usuń zakomentowany kod

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+            sumary = 0;
+            dummyPathString = "";
+        }
+//        for (String iter : treeSet) {
```
usuń zakomentowany kod

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+            }
+            if (sumary == n) {
+                treeSet.add(dummyPathString);
+//                allPasscount++;
```
usuń zakomentowany kod

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+    public static int getRandomNumber(int min, int max) {
+
+        ;
+//        System.out.print(k);
```
usuń zakomentowany kod

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+
+        Zad5 klasa = new Zad5();
+        System.out.println(klasa.climbStairshash(5)) ;
+//        System.out.println(klasa.climbStairs(3));
```
usuń zakomentowany kod

zimi-pl commented on src/test/java/pl/zimi/zadanka/Zad5Test.java:
```diff
@@ -0,0 +1,13 @@
+
+public class Zad5Test {
+    @Test
+    void n3 ()
```
tylko jeden test? reszta jest gdzieś indziej?

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+
+
+    public static int climbStairsTreeSet(int n) {
+        Set<String> treeSet = new TreeSet<>();
```
nazwa zmiennej -> to jasne że to treeset... pytanie po co ta kolekcja

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+        int allFilteredCount = 0;
+        for (Integer k = 0; k < 100000000; k++) {
+            while (sumary < n) {
+                randomPath = (int) ((Math.random() * (2)) + 1);
```
nie bardzo w sumie rozumiem logikę... to nie jest problem kodu tylko bardziej mój ale chyba będziemy musieli przegadać logikę tego kodu żebym zrozumiał

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+        int left=1;
+        int right=2;
+
+        while ((pointer!=0)&&(pointer!=key)){
```
to wygląda jak pętla nieskończona dla np. climbStairs(2)

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,119 @@
+
+    public static int climbStairsTreeSet(int n) {
+        Set<String> treeSet = new TreeSet<>();
+        int sumary = 0;
```
summary jako podsumowanie jest przez dwa m chyba że to co innego miało być

sm000k committed:
Zad 5 testy i rozwiązania. Napisałem 3 metody. Wszystkie metody dają poprawny wynik. Jednak tylko climbStairsQuicker przechodzi testy bez przekroczenia czasu na wykonanie.

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad3.java:
```diff
@@ -0,0 +1,26 @@
+        return false;
+    }
+
+    ;
```
niepotrzebny średnik

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+
+    // rozwiązanie poprawne ale wolne
+    public static int climbStairsTreeStyle(int key, int currentValue, int carry) {
+        if (currentValue == key) return ++carry;
```
nawiasy klamrowe + nowa linia trzymaj się konwencji

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+    // rozwiązanie poprawne ale wolne
+    public static int climbStairsTreeStyle(int key, int currentValue, int carry) {
+        if (currentValue == key) return ++carry;
+        if (currentValue > key) return 0;
```
j.w.

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+    public static int climbStairsTreeStyle(int key, int currentValue, int carry) {
+        if (currentValue == key) return ++carry;
+        if (currentValue > key) return 0;
+        else
```
zawsze klamry

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+    // rozwiązanie poprawne ale za wolne
+    public static int climbStairs(int key) {
+        if (key > 0) {
+            if (key == 1) return 1;
```
zawsze klamry

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+    public static int climbStairs(int key) {
+        if (key > 0) {
+            if (key == 1) return 1;
+            if (key == 2) return 2;
```
klamry

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+            if (key == 1) return 1;
+            if (key == 2) return 2;
+            return climbStairs(key - 2) + climbStairs(key - 1);
+        } else return 0;
```
klamry

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+
+    // rozwiazanie przyjete przez leetcode
+    public static int climbStairsQuicker(int key) {
+        if (key == 2) return 2;
```
klamry

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+    // rozwiazanie przyjete przez leetcode
+    public static int climbStairsQuicker(int key) {
+        if (key == 2) return 2;
+        if (key == 1) return 1;
```
klamry

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5Drafts.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.TreeSet;
+
+public class Zad5Drafts {
+    public static int climbStairsTreeSet(int n) {
```
czy ten kod chcesz zachować w jakimś celu? jeśli tak to spoko ale w taki sposób żeby jasne było po co chcesz go zachować (np. do nauki)

jeśli chcesz go zachować z sentymentu... to go usuń... zadaniem gita jest trzymać stare rzeczy... aktualna wersja powinna trzymać  tylko potrzebne...

sm000k replied:
z sentymentu, dużo się namęczyłem przy nim. Po przejrzeniu nie widzę w nim wartości edukacyjnych, więc usuwam.

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5Drafts.java:
```diff
@@ -0,0 +1,76 @@
+            }
+            if (sumary == n) {
+                treeSet.add(dummyPathString);
+//                allPasscount++;
```
usuń komentarz

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5Drafts.java:
```diff
@@ -0,0 +1,76 @@
+            sumary = 0;
+            dummyPathString = "";
+        }
+//        for (String iter : treeSet) {
```
usuń zakomentowany kod

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5Drafts.java:
```diff
@@ -0,0 +1,76 @@
+            }
+            if (sumary == n) {
+                treeSet.add(dummyPathString);
+//                allPasscount++;
```
komentarz

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5Drafts.java:
```diff
@@ -0,0 +1,76 @@
+            sumary = 0;
+            dummyPathString = "";
+        }
+//        for (String iter : treeSet) {
+////            System.out.println("allFilteredcount=" + allFilteredCount + " iter=" + iter);
+//            allFilteredCount++;
+//        }
```
komentarz

zimi-pl commented on src/test/java/pl/zimi/zadanka/Zad3Test.java:
```diff
@@ -0,0 +1,52 @@
+public class Zad3Test {
+    @Test
+    void checkZero(){
+        Zad3 object = new Zad3();
```
prosiłbym o porządek...

dbanie o wcięcia i nie przesadzanie z dodatkowymi liniami...
kod powinien wyglądać, że ktoś o niego dba...

firmy bardzo cenią sobie używanie konwencji i standardów

sm000k replied:
oczywiście, poprawiam

zimi-pl commented on src/test/java/pl/zimi/zadanka/Zad4Test.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    void checkstring() {
+
+        Zad4 object = new Zad4();
```
j.w.

sm000k replied:
ok

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,48 @@
+public class Zad5 {
+
+    // rozwiązanie poprawne ale wolne
+    public static int climbStairsTreeStyle(int key, int currentValue, int carry) {
```
generalnie nie mam nic do nazwy... ale nie bardzo wiem dlaczego TreeStyle... nie mówię że jest źle bo nie wiem czemu twierdzisz że to jest "tree style"

sm000k replied:
dwie równoległe rekurencje badające wszystkie rezultaty dwóch operacji skojarzyły mi się z drzewem binarny, gdzie liczba to węzeł, a operacja dodania 1 lub 2 to ścieżka do węzła. Nie poprawiam teraz nazwy póki nie ustalimy, czy moje skojarzenie jest mylne

zimi-pl replied:
zostałbym po prostu przy rekurencji: "RecurrenceStyle"
wywołania metod faktycznie układają się w drzewa (niekoniecznie binarne - jak lekko zmodyfikujesz problem)... ale to jest typowe dla rekurencji...

jak myślę o drzewach i używaniu tego pojęcia to myślę o drzewiastych strukturach danych... a takich tutaj brak... dlatego nie stosowałbym tej nazwy

lista, graf, drzewo, mapa to struktury danych nie używałbym nazw jeśli nie posiadamy odpowiedniej struktury

sm000k replied:
ok dotarło

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5Drafts.java:
```diff
@@ -0,0 +1,76 @@
+
+public class Zad5Drafts {
+    public static int climbStairsTreeSet(int n) {
+        Set<String> treeSet = new TreeSet<>();
```
wiemy że w tej zmiennej jest treeSet, taka nazwa nic nie wnosi... co ten set przechowuje biznesowo? innymi słowy co przechowuje ten set w kontekście rozwiązywanego problemu? dostosuj nazwę

sm000k committed:
Zad 5 poprawki wcięć komentarzy zbędnego kodu

sm000k reviewed (changes requested):
prosiłbym, o sugestie dotyczącą nazwy funkcji Treestyle pod twoim komentarzem do tej nazwy wyjasniłem powód. TreeStyle brzmi,źle. Może wtedy climbStairsBinaryTreeSolutionAlike

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,62 @@
+
+    // rozwiązanie poprawne ale za wolne
+    public static int climbStairs(int key) {
+        if (key > 0) {
+            if (key == 1) {
+                return 1;
+            }
+            if (key == 2) {
+                return 2;
+            }
```
kod jest poprawny ale można go uprościć
zauważ że kod dla key = 0 ma zwrócić 0
dla key = 1 -> 1
dla key = 2 -> 2
dla większych wartości zachowanie się zmienia

ponadto cechą dobrego kodu jest reakcja na inne dane...
co powinno się wydarzyć gdy za key dostaniesz ujemną wartość

sm000k replied:
zapropowanieł wywołanie wyjątku :
throw new IllegalArgumentException("input variable StairNumber cannot be less than 0");

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    // rozwiązanie poprawne ale za wolne
+    public static int climbStairs(int key) {
```
popraw nazwę "key" co biznesowo znaczy ta wartość

sm000k replied:
:D biznesowo to nic nie znaczy bo nie zarabia pieniędzy :)
tam gdzie "key" był uzywany w kontekście zmiennej ulegającej zmianie w trakcie kolejny rekurencji przemianowałem zmienną na stairNumber. W miejscach gdzie kontekst zmiennej key  wskazywał, że jej wartość będzie służyła do porównywania ze zmienną ulegającą zmianie w trakcie wykonywania metody przemianowałem key na lastStairNumber

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,62 @@
+
+    // rozwiazanie przyjete przez leetcode
+    public static int climbStairsQuicker(int key) {
+        if (key == 2) {
```
to samo co wcześniej

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad5.java:
```diff
@@ -0,0 +1,62 @@
+        int secondValue = 2;
+        int firstValue = 1;
+        int iterVal = 3;
+        while (iterVal++ <= key) {
```
wolałbym tutaj pętle for ona jest do takich iteracji 
while jest bardziej w celu takich nieprzewidywalnych pętli... gdzie z reguły jakimś boolean-em operujesz

while (temperature  < 0) {

while (x * y %10 != 0) {

zimi-pl replied:
ale nie jest to błędem... żeby to było jasne... chodzi bardziej o konwencje...

sm000k replied:
jasne nie zwróciłem uwagi, że jest nietypowo napisane

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad4.java:
```diff
@@ -0,0 +1,19 @@
+ */
+public class Zad4 {
+    public static String reverseString(String s) {
+        if (s == null) return null;
```
klamry

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad4.java:
```diff
@@ -0,0 +1,19 @@
+public class Zad4 {
+    public static String reverseString(String s) {
+        if (s == null) return null;
+        if (s.equals("")) return "";
```
klamry

zimi-pl commented on src/main/java/pl/zimi/zadanka/Zad4.java:
```diff
@@ -0,0 +1,19 @@
+        String[] words = s.split(" ");
+        String reversedString = "";
+        for (int i = words.length - 1; i >= 0; i--) {
+            if (!words[i].equals("")) reversedString += " " + words[i];
```
klamry

zimi-pl commented on src/test/java/pl/zimi/zadanka/Zad5Test.java:
```diff
@@ -0,0 +1,72 @@
+    @Test
+    void n3TreeVersion() {
+        Zad5 object = new Zad5();
+        Assertions.assertEquals(3, object.climbStairsTreeStyle(3, 0, 0));
```
metoda w testach powinna mieć tylko jeden parametr...

dwa kolejne parametry to jakieś internalowe rzeczy...
je chcemy ukrywać (enkapsulować) nie chcemy żeby ktoś tu mógł wrzucić -1

tą metodę możemy ukryć za pomocą private... żebyśmy tylko wewnątrz klasy mogli jej użyć

sm000k replied:
ok utworzyłem kolejną metodę typu public która wywołuje metodę właściwą, której typ zmieniłem na private.

sm000k committed:
Zad 5 zmiana nazwy key ; propozycja obsługi zmiennej wejsciowej mniejszej od zera ;
zmiana sposobu wywołania metody climbStairsRecurrencyStyle, by enkapsulować np wartości zmiennych inicjujących metody
